File: lc/coreprep/implementation/patterns_reference_implementation/two_pointers/DutchNationalFlag.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n) / O(1)
public class DutchNationalFlag {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        sort(a, 0, n - 1);

        for (int e : a) {
          System.out.printf("%d ", e);
        }
        System.out.println();
      }
    }
  }

  private static void sort(int[] a, int low, int high) {
    int i = 0;

    while (i <= high) {
      if (a[i] == 0) {
        swap(a, i, low);
        i++;
        low++;
      } else if (a[i] == 1) {
        i++;
      } else {
        swap(a, i, high);
        high--;
      }
    }
  }

  private static void swap(int[] a, int x, int y) {
    int t = a[x];
    a[x] = a[y];
    a[y] = t;
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/dutch_national_flag.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

void swap(vector<int> &a, int x, int y) {
  int t = a[x];
  a[x] = a[y];
  a[y] = t;
}

// O(n) / O(1)
void sort(vector<int> &a) {
  int low = 0, high = a.size() - 1;

  int i = 0;
  while (i <= high) {
    if (a[i] == 0) {
      swap(a, i, low);
      low++;
      i++;
    } else if (a[i] == 1) {
      i++;
    } else {
      swap(a, i, high);
      high--;
    }
  }
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    sort(a);

    for (int e : a) {
      cout << e << " ";
    }
    cout << "\n";
  }

  return 0;
}